neural_modelling/src/spike_source/poisson_variable/spike_source_poisson_variable.c

```diff
@@ -0,0 +1,728 @@
+    //////////////////////////////////////////////////////
+
+    address_t address = data_specification_get_data_address();
+
+    // clone read function and omit re-reading the starting parts. or split it, and
+    // have read parameters, and read poisson source structs functions. Both can be
+    // called at initialise, but only the spike source structs will be updated at resume?
+
+    if (!read_poisson_parameters(
+            data_specification_get_region(POISSON_PARAMS, address))){
+        log_error("failed to reread the poisson parameters from SDRAM");
+        rt_error(RTE_SWERR);
+    }
+
+    // time_to_change = 0;
+
+    // print spike sources for debug purposes
+    print_spike_sources();
+}
+
+//! \brief stores the poisson parameters back into sdram for reading by the
+//! host when needed
+//! \return None
+bool store_poisson_parameters(){
+    log_debug("stored_parameters: starting");
+
+    // Get the address this core's DTCM data starts at from SRAM
+    address_t address = data_specification_get_data_address();
+    address = data_specification_get_region(POISSON_PARAMS, address);
+    uint32_t seed_size = sizeof(mars_kiss64_seed_t) / sizeof(uint32_t);
+
+    num_spike_sources = address[PARAMETER_SEED_START_POSITION + seed_size];
+    log_debug("\t spike sources = %u", num_spike_sources);
+
+    // store array of spike sources into sdram for reading by the host
+    if (num_spike_sources > 0) {
+	// uint32_t spikes_offset = PARAMETER_SEED_START_POSITION + seed_size + 4;
+	// memcpy(
+	//     &address[spikes_offset], spike_source_array,
+	//     num_spike_sources * sizeof(spike_source_t));
+	// // store spike source data into DTCM
+	// uint32_t spikes_offset = PARAMETER_SEED_START_POSITION + seed_size + 6;
+        memcpy(
+            &address[_calc_memory_loc()], spike_source_array,
+            num_spike_sources * sizeof(spike_source_t));
+    }
+
+    log_debug("stored_parameters : completed successfully");
+    return true;
+}
+
+//! \brief handles spreading of poisson spikes for even packet reception at
+//! destination
+//! \param[in] spike_key: the key to transmit
+//! \return None
+void _send_spike(uint spike_key) {
+
+    // Wait until the expected time to send
+    while (tc[T1_COUNT] > expected_time) {
+
+        // Do Nothing
+    }
+    expected_time -= time_between_spikes;
+
+    // Send the spike
+    log_debug("Sending spike packet %x at %d\n", spike_key, time);
+    while (!spin1_send_mc_packet(spike_key, 0, NO_PAYLOAD)) {
+        spin1_delay_us(1);
+    }
+}
+
+//! \brief records spikes as needed
+//! \param[in] neuron_id: the neurons to store spikes from
+//! \param[in] n_spikes: the number of times this neuron has spiked
+//!
+static inline void _mark_spike(uint32_t neuron_id, uint32_t n_spikes) {
+    if (recording_flags > 0) {
+        if (n_spike_buffers_allocated < n_spikes) {
+            uint32_t new_size = 8 + (n_spikes * spike_buffer_size);
```

Comment:
These sorts of size calculations are a great time to use a small macro:

    #define SPIKE_COLLECTION_SIZE(n_spikes) (8 + ((n_spikes) * spike_buffer_size))

    new_spikes = spin1_malloc(SPIKE_COLLECTION_SIZE(n_spikes));

etc. Also, there's a magic constant (`8`) here; it's usually better if those can be named or calculated from something else (e.g., `sizeof(some_prefix_struct)`).

Reply:
@oliverrhodes

Reply:
Not a blocker anyway.

Reply:
code optimised now!